staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -143,107 +140,166 @@ func (h *peerProxyHandler) WrapHandler(handler http.Handler) http.Handler {
 			gvr.Group = "core"
 		}
 
-		apiservers, err := h.findServiceableByServers(gvr)
-		if err != nil {
-			// resource wasn't found in SV informer cache which means that resource is an aggregated API
-			// or a CR. This situation is ok to be handled by local handler.
+		if h.shouldServeLocally(gvr) {
 			handler.ServeHTTP(w, r)
 			return
 		}
 
-		locallyServiceable, peerEndpoints, err := h.resolveServingLocation(apiservers)
+		// find servers that are capable of serving this request
+		peerServerIDs := h.findServiceableByPeerFromAggDiscoveryCache(gvr)
+		if len(peerServerIDs) == 0 {
+			klog.Errorf("gvr %v is not served by anything in this cluster", gvr)
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		peerEndpoints, err := h.resolveServingLocation(peerServerIDs)
 		if err != nil {
 			gv := schema.GroupVersion{Group: gvr.Group, Version: gvr.Version}
 			klog.ErrorS(err, "error finding serviceable-by apiservers for the requested resource", "gvr", gvr)
 			responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable("Error getting ip and port info of the remote server while proxying"), h.serializer, gv, w, r)
 			return
 		}
 
-		// pass request to the next handler if found the gvr locally.
-		// TODO: maintain locally serviceable GVRs somewhere so that we dont have to
-		// consult the storageversion-informed map for those
-		if locallyServiceable {
-			handler.ServeHTTP(w, r)
-			return
+		rand := rand.Intn(len(peerEndpoints))
+		peerEndpoint := peerEndpoints[rand]
+		h.proxyRequestToDestinationAPIServer(r, w, peerEndpoint)
+	})
+}
+
+func (h *peerProxyHandler) startLocalDiscoveryCacheInvalidation(stopCh <-chan struct{}) {
```

Comment:
can you doc document what is this doing and why?
trying to figure out from the code is hard without the entire context of the peerproxy

Reply:
Added comments